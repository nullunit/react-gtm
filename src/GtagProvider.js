import React from 'react';
import PropTypes from 'prop-types';
import GtagScript from './GtagScript';
import GtagManagerScript from './GtagManagerScript';
import { getGtagContext } from './GtagContext';

export default class GtagProvider extends React.Component {

    static defaultProps = {
        id: null,
        useTagManager: false,
        appName: null,
        appVersion: null
    };

    static propTypes = {
        id: PropTypes.string.isRequired,
        useTagManager: PropTypes.bool,
        appName: PropTypes.string,
        appVersion: PropTypes.string,
        children: PropTypes.node.isRequired
    };

    constructor(props) {
        super(props);
        this._createDataLayer();
    }

    _createDataLayer() {
        window.dataLayer = window.dataLayer || [];
    }

    _renderChild(context) {
        const child = React.Children.only(this.props.children);
        return React.cloneElement(child, { gtag: context }); 
    }
    
    render() {                                   
        const context = getGtagContext(window, this.props.id);
        context.appName = this.props.appName;
        context.appVersion = this.props.appVersion;
                                                                        
        if (this.props.useTagManager) { 
            return (<GtagManagerScript id={ this.props.id } context={ context }>{ this._renderChild(context) }</GtagManagerScript>);
        }
        else {                    
            return (<GtagScript id={ this.props.id } context={ context }>{ this._renderChild(context) }</GtagScript>);
        }
    }
}
