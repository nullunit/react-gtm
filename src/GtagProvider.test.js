import React from 'react';
import renderer from 'react-test-renderer';
import GtagProvider from './GtagProvider';
jest.mock('./GtagManagerScript');
jest.mock('./GtagScript');

const GTM_ID = 'GTM-123456';

describe('GtagProvider tests', () => {

    it('Create GtagProvider for Google Analytics', () => {

        const props = {
            id: GTM_ID,
            useTagManager: false
        };

        //test render snapshot
        const json = renderer
            .create(<GtagProvider { ...props }><h1>mock child</h1></GtagProvider>)
            .toJSON();

        expect(json).toMatchSnapshot();
    });

    it('Create GtagProvider for Google Tag Manager', () => {

        const props = {
            id: GTM_ID,
            useTagManager: true
        };

        //test render snapshot
        const json = renderer
            .create(<GtagProvider { ...props }><h1>mock child</h1></GtagProvider>)
            .toJSON();

        expect(json).toMatchSnapshot();
    });
});
