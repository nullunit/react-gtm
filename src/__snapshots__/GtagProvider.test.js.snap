// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GtagProvider tests Create GtagProvider for Google Analytics 1`] = `
<h1>
  mock render Google Analytics script to page
</h1>
`;

exports[`GtagProvider tests Create GtagProvider for Google Tag Manager 1`] = `
<h1>
  mock render Google Tag Manager script to page
</h1>
`;
